Feature | integrate and improve automated validation tests script

## What?

- ✅ Include and improve validation tests [Leverage reference architecture script](https://github.com/binbashar/le-tf-infra-aws/blob/master/%40bin/scripts/validate-terraform-layout.sh)

- **Possible Checks** 
  - terraform fmt
  - terraform tflint
  - terraform validate
  - security (https://github.com/binbashar/le-tf-infra-aws/issues/73)
  - networking (overlapping, other validations?)

## Why?

- ✅ Validate that the terraform code canonical format, layout and security enforcing best practices. 
- ✅ Early phase error detection.


## Reference 

Current version is functional and could be found at the @bin folder in the Terraform Ref Arch repo.

```
╭─    ~/Binbash/repos/Leverage/ref-architecture/le-tf-infra-aws  on   master ····· ✔  at 12:32:08 
╰─ cd shared/base-network

╭─    ~/B/r/L/ref-architecture/le-tf-infra-aws/shared/base-network  on   master ·· ✔  at 12:32:13 
╰─ leverage validate_layout
[DEBUG] Found config file: /Users/exequielbarrirero/Binbash/repos/Leverage/ref-architecture/le-tf-infra-aws/shared/base-network/build.env
[DEBUG] Found config file: /Users/exequielbarrirero/Binbash/repos/Leverage/ref-architecture/le-tf-infra-aws/build.env
[ build.py - Starting task "_checkdir" ]
[ build.py - Completed task "_checkdir" ]
[ build.py - Starting task "validate_layout" ]
[INFO] CWD=/Users/exequielbarrirero/Binbash/repos/Leverage/ref-architecture/le-tf-infra-aws/shared/base-network
[INFO] REPO_DIR=/Users/exequielbarrirero/Binbash/repos/Leverage/ref-architecture/le-tf-infra-aws
[INFO] ACCOUNT_NAME: shared
[INFO] CATEGORY_NAME: base-network

[INFO] Checking if backend key starts with shared ...
[INFO]  -> OK
[INFO] Checking if backend key starts with shared and follows with base-network ...
[ERROR]  -> FAILED
[INFO] Checking if backend.config profile contains shared ...
[INFO]  -> OK
[INFO] Checking if backend.config bucket contains shared ...
[INFO]  -> OK
[INFO] Checking if backend.config dynamodb_table contains shared ...
[INFO]  -> OK
[INFO] Checking if base.config  contains shared ...
[INFO]  -> OK
[ build.py - Completed task "validate_layout" ]
```

--- a/leverage/container.py
+++ b/leverage/container.py
@@ -450,7 +450,7 @@ class TerraformContainer(LeverageContainer):
 
         logger.debug(f"[bold cyan]Running with entrypoint:[/bold cyan] {self.entrypoint}")
 
-    def _check_for_layer_location(self):
+    def check_for_layer_location(self):
         """ Make sure the command is being ran at layer level. If not, bail. """
         if self.cwd in (self.common_config_dir, self.account_config_dir):
             logger.error("Currently in a configuration directory, no Terraform command can be run here.")
@@ -472,7 +472,7 @@ class TerraformContainer(LeverageContainer):
 
     def start_in_layer(self, command, *arguments):
         """ Run a command that can only be performed in layer level. """
-        self._check_for_layer_location()
+        self.check_for_layer_location()
 
         return self.start(command, *arguments)
 

--- a/leverage/modules/terraform.py
+++ b/leverage/modules/terraform.py
@@ -1,11 +1,16 @@
+import re
+import hcl2
 import click
 from click.exceptions import Exit
 
+from leverage import logger
 from leverage._internals import pass_state
 from leverage._internals import pass_container
 from leverage.container import get_docker_client
 from leverage.container import TerraformContainer
 
+REGION = (r"global|(?:[a-z]{2}-(?:gov-)?"
+          r"(?:central|north|south|east|west|northeast|northwest|southeast|southwest|secret|topsecret)-[1-4])")
 
 @click.group()
 @pass_state
@@ -20,7 +25,9 @@ def terraform(state):
     state.container.ensure_image()
 
 
-CONTEXT_SETTINGS = {"ignore_unknown_options": True}
+CONTEXT_SETTINGS = {
+    "ignore_unknown_options": True
+}
 
 
 @terraform.command(context_settings=CONTEXT_SETTINGS)
@@ -28,8 +35,11 @@ CONTEXT_SETTINGS = {"ignore_unknown_options": True}
               is_flag=True)
 @click.argument("args", nargs=-1)
 @pass_container
-def init(tf, no_backend, args):
+@click.pass_context
+def init(context, tf, no_backend, args):
     """ Initialize this layer. """
+    context.invoke(validate_layout) # Validate layout before attempting to initialize Terraform
+
     backend_config = ["-backend=false" if no_backend else f"-backend-config={tf.TF_BACKEND_TFVARS}"]
     args = backend_config + list(args)
     exit_code = tf.start_in_layer("init", *args)
@@ -109,12 +119,12 @@ def shell(tf, mfa, sso):
     tf.start_shell()
 
 
-@terraform.command("format")
+@terraform.command()
 @click.option("--check",
               is_flag=True,
               help="Only perform format checking, do not rewrite the files.")
 @pass_container
-def _format(tf, check):
+def fmt(tf, check):
     """ Check if all files meet the canonical format and rewrite them accordingly. """
     args = ["-recursive"]
     if check:
@@ -130,6 +140,83 @@ def validate(tf):
     """ Validate code of the current directory. Previous initialization might be needed. """
     tf.disable_authentication()
     tf.start("validate")
+
+
+def _make_layer_backend_key(cwd, account_dir, account_name):
+    """ Create expected backend key.
+
+    Args:
+        cwd (pathlib.Path): Current Working Directory (Layer Directory)
+        account_dir (pathlib.Path): Account Directory
+        account_name (str): Account Name
+
+    Returns:
+        list: Backend bucket key parts
+    """
+    layer_path = cwd.relative_to(account_dir)
+    layer_path = layer_path.as_posix().split("/")    
+    # Remove region directory
+    layer_path = layer_path[1:] if re.match(REGION, layer_path[0]) else layer_path
+    # Remove layer name prefix
+    layer_name_parts = layer_path[0].split("-")
+    layer_name_parts = layer_name_parts[1:] if layer_name_parts[0] in ("base", "tools") else layer_name_parts
+    layer_path[0] = "-".join(layer_name_parts)
+    return [account_name, *layer_path]
+
+
+@terraform.command("validate-layout")
+@pass_container
+def validate_layout(tf):
+    """ Validate layer conforms to Leverage convention. """
+    tf.check_for_layer_location()
+
+    # Check for `environment = <account name>` in account.tfvars
+    account_name = tf.account_conf.get("environment")
+    logger.info("Checking environment name definition in [bold]account.tfvars[/bold]...")
+    if account_name is None:
+        logger.error("[red]✘ FAILED[/red]\n")
+        raise Exit(1)
+    logger.info("[green]✔ OK[/green]\n")
+
+    # Check if account directory name matches with environment name
+    if tf.account_dir.stem != account_name:
+        logger.warning("[yellow]‼[/yellow] Account directory name does not match environment name.\n"
+                       f"  Expected [bold]{account_name}[/bold], found [bold]{tf.account_dir.stem}[/bold]\n")
+
+    config_tf = tf.cwd / "config.tf"
+    try:
+        config_tf = hcl2.loads(config_tf.read_text()) if config_tf.exists() else {}
+        backend_key = config_tf["terraform"][0]["backend"][0]["s3"]["key"].split("/")
+    except (KeyError, IndexError):
+        logger.error("[red]✘[/red] Malformed [bold]config.tf[/bold] file. Missing Terraform backend bucket key.")
+        raise Exit(1)
+    except:
+        logger.error("[red]✘[/red] Malformed [bold]config.tf[/bold] file. Unable to parse.")
+        raise Exit(1)
+
+    # Check backend bucket key
+    expected_backend_key = _make_layer_backend_key(tf.cwd, tf.account_dir, account_name)
+    logger.info(f"Checking if backend key matches '{'/'.join(expected_backend_key)}/terraform.tfstate'...")
+    if backend_key[:-1] == expected_backend_key:
+        logger.info("[green]✔ OK[/green]\n")
+    else:
+        logger.error("[red]✘ FAILED[/red]\n")
+
+    backend_tfvars = tf.account_config_dir / tf.BACKEND_TFVARS
+    backend_tfvars = hcl2.loads(backend_tfvars.read_text()) if backend_tfvars.exists() else {}
+
+    logger.info("Checking [bold]backend.tfvars[/bold]:\n")
+    names_prefix = f"{tf.project}-{account_name}"
+
+    # Check profile, bucket and dynamo table names:
+    for field in ("profile", "bucket", "dynamodb_table"):
+        logger.info(f"Checking if {field.replace('_', ' ')} starts with {names_prefix}...")
+        if backend_tfvars.get(field, "").startswith(names_prefix):
+            logger.info("[green]✔ OK[/green]\n")
+        else:
+            logger.error("[red]✘ FAILED[/red]\n")
+
+    logger.info("Done.")
 
 
 @terraform.command("import")

--- a/leverage/modules/terraform.py
+++ b/leverage/modules/terraform.py
@@ -119,12 +119,12 @@ def shell(tf, mfa, sso):
     tf.start_shell()
 
 
-@terraform.command()
+@terraform.command("format")
 @click.option("--check",
               is_flag=True,
               help="Only perform format checking, do not rewrite the files.")
 @pass_container
-def fmt(tf, check):
+def _format(tf, check):
     """ Check if all files meet the canonical format and rewrite them accordingly. """
     args = ["-recursive"]
     if check:

--- a/leverage/modules/terraform.py
+++ b/leverage/modules/terraform.py
@@ -38,7 +38,10 @@ CONTEXT_SETTINGS = {
 @click.pass_context
 def init(context, tf, no_backend, args):
     """ Initialize this layer. """
-    context.invoke(validate_layout) # Validate layout before attempting to initialize Terraform
+    is_layout_valid = context.invoke(validate_layout) # Validate layout before attempting to initialize Terraform
+    if not is_layout_valid:
+        logger.error("Layer configuration is not valid. Exiting.")
+        raise Exit(1)
 
     backend_config = ["-backend=false" if no_backend else f"-backend-config={tf.TF_BACKEND_TFVARS}"]
     args = backend_config + list(args)
@@ -154,7 +157,7 @@ def _make_layer_backend_key(cwd, account_dir, account_name):
         list: Backend bucket key parts
     """
     layer_path = cwd.relative_to(account_dir)
-    layer_path = layer_path.as_posix().split("/")    
+    layer_path = layer_path.as_posix().split("/")
     # Remove region directory
     layer_path = layer_path[1:] if re.match(REGION, layer_path[0]) else layer_path
     # Remove layer name prefix
@@ -194,6 +197,9 @@ def validate_layout(tf):
         logger.error("[red]✘[/red] Malformed [bold]config.tf[/bold] file. Unable to parse.")
         raise Exit(1)
 
+    # Flag to report layout validity
+    valid_layout = True
+
     # Check backend bucket key
     expected_backend_key = _make_layer_backend_key(tf.cwd, tf.account_dir, account_name)
     logger.info(f"Checking if backend key matches '{'/'.join(expected_backend_key)}/terraform.tfstate'...")
@@ -201,6 +207,7 @@ def validate_layout(tf):
         logger.info("[green]✔ OK[/green]\n")
     else:
         logger.error("[red]✘ FAILED[/red]\n")
+        valid_layout = False
 
     backend_tfvars = tf.account_config_dir / tf.BACKEND_TFVARS
     backend_tfvars = hcl2.loads(backend_tfvars.read_text()) if backend_tfvars.exists() else {}
@@ -215,8 +222,9 @@ def validate_layout(tf):
             logger.info("[green]✔ OK[/green]\n")
         else:
             logger.error("[red]✘ FAILED[/red]\n")
+            valid_layout = False
 
-    logger.info("Done.")
+    return valid_layout
 
 
 @terraform.command("import")

